File: src/itkach/aard2/lookup/LookupFragment.java
package itkach.aard2.lookup;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModelProvider;

import java.util.Timer;
import java.util.TimerTask;

import itkach.aard2.Application;
import itkach.aard2.BaseListFragment;
import itkach.aard2.MainActivity;
import itkach.aard2.R;
import itkach.aard2.SlobHelper;
import itkach.aard2.article.ArticleCollectionActivity;
import itkach.aard2.prefs.AppPrefs;
import itkach.aard2.utils.ClipboardUtils;
import itkach.slob.Slob;

public class LookupFragment extends BaseListFragment implements LookupListener, SearchView.OnQueryTextListener {
    private final static String TAG = LookupFragment.class.getSimpleName();

    private Timer timer;
    private SearchView searchView;
    private LookupResultAdapter listAdapter;
    private LookupViewModel viewModel;

    @Override
    protected int getEmptyIcon() {
        return R.drawable.ic_search;
    }

    @Override
    protected CharSequence getEmptyText() {
        return "";
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        viewModel = new ViewModelProvider(this).get(LookupViewModel.class);
        Application app = (Application) requireActivity().getApplication();
        app.addLookupListener(this);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setBusy(false);
        listAdapter = new LookupResultAdapter(SlobHelper.getInstance().lastLookupResult);
        recyclerView.setAdapter(listAdapter);
    }

    @Override
    public void onResume() {
        super.onResume();
        FragmentActivity activity = requireActivity();
        if (activity instanceof MainActivity) {
            ((MainActivity)activity).requireActionBar().setTitle(null);
            ((MainActivity)activity).requireActionBar().setSubtitle(R.string.subtitle_lookup);
            if (AppPrefs.disableRandomLookup()) {
                ((MainActivity) activity).hideFab();
            } else {
                ((MainActivity) activity).displayFab(R.drawable.ic_auto_awesome, R.string.action_open_random_article, v -> {
                    Slob.Blob blob = SlobHelper.getInstance().findRandom();
                    if (blob == null) {
                        Toast.makeText(activity, R.string.article_collection_nothing_found, Toast.LENGTH_SHORT).show();
                        return;
                    }
                    Intent intent = new Intent(activity, ArticleCollectionActivity.class);
                    intent.setData(SlobHelper.getInstance().getHttpUri(blob));
                    startActivity(intent);
                });
            }
        }
    }

    @Override
    public void onPause() {
        FragmentActivity activity = requireActivity();
        if (activity instanceof MainActivity) {
            ((MainActivity) activity).hideFab();
        }
        super.onPause();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        timer = new Timer();
        inflater.inflate(R.menu.lookup, menu);
        MenuItem lookupMenu = menu.findItem(R.id.action_lookup);
        View filterActionView = lookupMenu.getActionView();
        searchView = filterActionView.findViewById(R.id.search);
        searchView.setQueryHint(lookupMenu.getTitle());
        searchView.setIconified(false);
        searchView.setOnQueryTextListener(this);
        searchView.setOnCloseListener(() -> true);
        searchView.setSubmitButtonEnabled(false);
        if (!AppPrefs.showKeyboarOnLookup()) {
            searchView.clearFocus();
        }
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {
        super.onPrepareOptionsMenu(menu);
        if (AppPrefs.autoPasteInLookup()) {
            CharSequence clipboard = ClipboardUtils.take(requireContext());
            if (clipboard != null && viewModel != null) {
                searchView.setQuery(clipboard.toString(), false);
                viewModel.lookup(clipboard.toString());
                return;
            }
        }
        String query = AppPrefs.getLastQuery();
        searchView.setQuery(query, false);
        if (viewModel != null) {
            viewModel.lookupLastQuery();
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (searchView != null) {
            String query = searchView.getQuery().toString();
            outState.putString("lookupQuery", query);
        }
    }

    private void setBusy(boolean busy) {
        if (!busy) {
            TextView emptyText = emptyView.findViewById(R.id.empty_text);
            String msg = "";
            String query = AppPrefs.getLastQuery();
            if (!query.isEmpty()) {
                msg = getString(R.string.lookup_nothing_found);
            }
            emptyText.setText(msg);
        }
    }

    @Override
    public void onDestroy() {
        listAdapter = null;
        Application app = (Application) requireActivity().getApplication();
        app.removeLookupListener(this);
        super.onDestroy();
    }

    @Override
    public void onLookupStarted(String query) {
        setBusy(true);
    }

    @Override
    public void onLookupFinished(String query) {
        setBusy(false);
    }

    @Override
    public void onLookupCanceled(String query) {
        setBusy(false);
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }

    TimerTask scheduledLookup = null;
    @Override
    public boolean onQueryTextChange(String newText) {
//        Log.d(TAG, "new query text: " + newText);
        TimerTask doLookup = new TimerTask() {
            @Override
            public void run() {
                final String query = searchView.getQuery().toString();
                if (viewModel != null) {
                    viewModel.lookup(query);
                }
                scheduledLookup = null;
            }
        };
        final String query = searchView.getQuery().toString();
        if (!AppPrefs.getLastQuery().equals(query)) {
            if (scheduledLookup != null) {
                scheduledLookup.cancel();
            }
            scheduledLookup = doLookup;
            timer.schedule(doLookup, 500);
        }
        return true;
    }
}

File: src/itkach/aard2/LookupFragment.kt
package itkach.aard2

import android.content.Intent
import android.database.DataSetObserver
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.widget.AdapterView.OnItemClickListener
import android.widget.TextView
import androidx.appcompat.widget.SearchView
import java.util.Timer
import java.util.TimerTask

class LookupFragment : BaseListFragment(), LookupListener {
    private var timer: Timer? = null
    private var searchView: SearchView? = null
    private var app: Application? = null
    private var queryTextListener: SearchView.OnQueryTextListener? = null
    private var closeListener: SearchView.OnCloseListener? = null
    private lateinit var observer: DataSetObserver


    override val emptyIcon: Int
        get() = R.drawable.search_24px

    override val emptyText: CharSequence?
        get() = ""


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        app = requireActivity().application as Application
        app!!.addLookupListener(this)
    }

    override fun supportsSelection(): Boolean {
        return false
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setBusy(false)
        val listView = listView
        setListShown(true)
        listView.onItemClickListener = OnItemClickListener { parent, view, position, id ->
            Log.i("--", "Item clicked: $position")
            val intent = Intent(
                activity,
                ArticleCollectionActivity::class.java
            )
            intent.putExtra("position", position)
            startActivity(intent)
        }
        val app = requireActivity().application as Application
        listView.adapter = app.lastResult
//        this.observer = object : DataSetObserver() {
//            override fun onChanged() {
//                getListView().invalidateViews()
//            }
//        }
//        app.lastResult!!.registerDataSetObserver(observer)

        closeListener = SearchView.OnCloseListener { true }

        queryTextListener = object : SearchView.OnQueryTextListener {
            var scheduledLookup: TimerTask? = null

            override fun onQueryTextSubmit(query: String): Boolean {
                Log.d(TAG, "query text submit: $query")
                onQueryTextChange(query)
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                Log.d(TAG, "new query text: $newText")
                val doLookup: TimerTask = object : TimerTask() {
                    override fun run() {
                        val query = searchView!!.query.toString()
                        if (app.lookupQuery == query) {
                            return
                        }
                        activity!!.runOnUiThread { app.lookup(query) }
                        scheduledLookup = null
                    }
                }
                val query = searchView!!.query.toString()
                if (app.lookupQuery != query) {
                    if (scheduledLookup != null) {
                        scheduledLookup!!.cancel()
                    }
                    scheduledLookup = doLookup
                    timer!!.schedule(doLookup, 600)
                }
                return true
            }
        }
    }


    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        timer = Timer()
        inflater.inflate(R.menu.lookup, menu)
        val miFilter = menu.findItem(R.id.fldLookup)
        searchView = miFilter.actionView as SearchView?
        searchView!!.queryHint = miFilter.title
        searchView!!.isIconified = false
        searchView!!.setOnQueryTextListener(queryTextListener)
        searchView!!.setOnCloseListener(closeListener)
        searchView!!.isSubmitButtonEnabled = false
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        if (app!!.autoPaste()) {
            val clipboard = Clipboard.take(requireActivity())
            if (clipboard != null) {
                app!!.lookup(clipboard.toString(), false)
            }
        }
        val query: CharSequence = app!!.lookupQuery
        searchView!!.setQuery(query, true)
        if (app!!.lastResult!!.count > 0) {
            searchView!!.clearFocus()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (searchView != null) {
            val query = searchView!!.query.toString()
            outState.putString("lookupQuery", query)
        }
    }

    private fun setBusy(busy: Boolean) {
        if (view == null) {
            return
        }
        if (!busy) {
            val emptyText = (emptyView.findViewById<View>(R.id.empty_text) as TextView)
            var msg = ""
            val query = app!!.lookupQuery
            if (query != null && query.toString() != "") {
                msg = getString(R.string.lookup_nothing_found)
            }
            emptyText.text = msg
        }
    }

    override fun onDestroy() {
        if (timer != null) {
            timer!!.cancel()
        }
        val app = requireActivity().application as Application
        app.removeLookupListener(this)
        super.onDestroy()
    }

    override fun onLookupStarted(query: String?) {
        setBusy(true)
    }

    override fun onLookupFinished(query: String?) {
        setBusy(false)
    }

    override fun onLookupCanceled(query: String?) {
        setBusy(false)
    }

    companion object {
        private val TAG: String = LookupFragment::class.java.simpleName
    }
}
